### PR TITLE
Add MicroByte | Understanding JWT

### LINKED ISSUE #1: Add MicroByte for JWT
**MicroByte Title:**: Understanding JWT

**Domain:** 
- Security
- Auth

**Scope of learning for learners:** 
- Understand working and importance of JWT
- Learn to parse a JWT on their own

**Activities:** 
- Parse JWT to get the message, 
- Try Modifying JWT to show that tampering voids the JWT
- Create own JWT with a custom message

### PR DESCRIPTION
**Pull request by DesignrKnight**
### Checklist:
- [x] Added activities for learners
- [x] Didn't give the solutions to the added activities straight away
- [x] Followed a LearnByDoing approach
- [x] Clean and well documented code (if source code files are involved) 
- [x] Added images for better understanding if deemed necessary

### CODE REVIEW

**Review by sudhanshutiwari264 — APPROVED** [review]
I really liked both  creativity and content 👍
Let's wait till other Advocates review it !
great work man 🎉🎉
Keep contributing !

File: understanding-jwt/README.md
Comment (YourAKShaw) [inline understanding-jwt/README.md]: It seems like you were trying to bold the above three words. It's not working because of the extra space you have after the ** at the beginning.

File: understanding-jwt/README.md
Comment (YourAKShaw) [inline understanding-jwt/README.md]: The above two questions like where and why JWT is used can be left out for learners to explore. Kindly don't give out more info than needed for a MicroByte, a MicroByte should engage one with activities, it's not to be a read through doc.

Comment (YourAKShaw) [reply understanding-jwt/README.md]: If you prefer adding collapsible markdown instead of removing the answers, you can do that too.

File: understanding-jwt/README.md
Comment (YourAKShaw) [inline understanding-jwt/README.md]: `If you do make your program, do send a pull request`

This will give a nice opportunity to create a learner issue once the PR is merged. Just to give pointers :smiley:

**Review by YourAKShaw — CHANGES_REQUESTED** [review]
@DesignrKnight It seems like your MicroByte can do a little more work in the way you layout the content. I mean to say that there should be more interaction and less of just going through the doc. Kindly don't get me wrong. What you've come up with is of real fine quality, but can use some fine-tuning. Like the intro needn't have that much of theory, you can leave some blanks to be filled by the learners. Kindly make a few adjustments, whatever you think you can improve upon. I'm pretty sure the other reviewers can give more specifics on the same. Also, kindly get to these mild issues.

**Review by Kevinpaulose05 — CHANGES_REQUESTED** [review]
As of now, the content is good but it has some voids that need to be addressed and filled.

1.Give a mention of pre-requisites, like basic authorisation/back-end dev. Else ppl will be able to relate less
2.There is no mention of where JWT is produced - client-side or server-side. And what happens from that side to the other side. Give an example of how it can be produced. Like a nodejs example of producing it and using it briefly.
3. There is a brief mention of sha256 algorithm. Explain encoding is done with base64 and before that the hashing algo is sha256(famous) or some other algo.
4..Give a practical usage example, like google apis. 
4. JWT is digitally signed, but more importantly it is cryptographically signed(but not encryted). Mention this if u agree with me
5. If user data is stored in jwt is HTTP or https safer? - can use this as a question

U have addressed the same issues like what/where/why jwt in multiple places. Address them in one place so as to cut short the theory.
Maybe u can show an API testing software's(like postman) screenshot, which include all the steps of getting the token, login, where the bearer token is put, what happens if it is not put.

Understanding JWT plainly is of no use and for a newbie won't make any sense. I can refer to this as a material to refresh my jwt knowledge but for a novice it will be impossible to imagine its actual use.

I agree this is are some pretty big tasks but doing these will only help u get the best out ur topic. Feel free to counter and discuss.

Cheers

**Comment by DesignrKnight** [comment]
Thanks for all the feedback. Doing the changes to make it better

**Comment by DesignrKnight** [comment]
Proposed Plan of Change
- [x] Adding pre-requisite
- [x] Removing long theory section and make it interactive
- [x] A live demo of client-side getting token and using it with API call. Will add at end of the activity, before the quiz
- [x] Additional Info on hash/digital sign
- [ ] Storing JWT and transmitting JWT. Best practices.
- [x] I'll try to make a server-side demo of JWT creation. Will add this also in the demo section with client-side demo

Does this sound good @Ak-Shaw @Kevinpaulose05

**Comment by Kevinpaulose05** [comment]
> Proposed Plan of Change
> 
> 1. Adding pre-requisite
> 2. Removing long theory section and make it interactive
> 3. A live demo of client-side getting token and using it with API call. Will add at end of the activity, before the quiz
> 4. Additional Info on hash/digital sign
> 5. Storing JWT and transmitting JWT. Best practices.
> 6. I'll try to make a server-side demo of JWT creation. Will add this also in the demo section with client-side demo
> 
> Does this sound good @Ak-Shaw @Kevinpaulose05

Spot on.

**Comment by YourAKShaw** [comment]
> Proposed Plan of Change
> 
>     1. Adding pre-requisite
> 
>     2. Removing long theory section and make it interactive
> 
>     3. A live demo of client-side getting token and using it with API call. Will add at end of the activity, before the quiz
> 
>     4. Additional Info on hash/digital sign
> 
>     5. Storing JWT and transmitting JWT. Best practices.
> 
>     6. I'll try to make a server-side demo of JWT creation. Will add this also in the demo section with client-side demo
> 
> 
> Does this sound good @Ak-Shaw @Kevinpaulose05

:100: :100:

File: understanding-jwt/README.md
Comment (DesignrKnight) [reply understanding-jwt/README.md]: Fixed in updated commits

File: understanding-jwt/README.md
Comment (DesignrKnight) [reply understanding-jwt/README.md]: Fixed in updated commits

File: understanding-jwt/README.md
Comment (DesignrKnight) [reply understanding-jwt/README.md]: Will add issues after PR merge

**Comment by DesignrKnight** [comment]
> Proposed Plan of Change
> 
> * [x]  Adding pre-requisite
> * [x]  Removing long theory section and make it interactive
> * [x]  A live demo of client-side getting token and using it with API call. Will add at end of the activity, before the quiz
> * [x]  Additional Info on hash/digital sign
> * [ ]  Storing JWT and transmitting JWT. Best practices.
> * [x]  I'll try to make a server-side demo of JWT creation. Will add this also in the demo section with client-side demo
> 
> Does this sound good @Ak-Shaw @Kevinpaulose05

Have done almost all the changes. Have not done JWT storing and transmitting section as that in itself is a vast topic

**Comment by Kevinpaulose05** [comment]
Put the "Understand how Signature works" after "Activity: Encode JWT". I think it makes more sense there

**Comment by DesignrKnight** [comment]
> Put the "Understand how Signature works" after "Activity: Encode JWT". I think it makes more sense there

Done

**Comment by Kevinpaulose05** [comment]
What is the industry-accepted pronunciation of JWT?
I think it is inappropriate to label it as "industry-accepted". U can rephrase it as it is commonly pronounced as "jot". I personally don't use that, but some do. So I have some second thoughts there.

**Comment by Kevinpaulose05** [comment]
> > Put the "Understand how Signature works" after "Activity: Encode JWT". I think it makes more sense there
> 
> Done

I am sorry. I meant before "Encode activity". Coz signature mention is there

**Comment by Kevinpaulose05** [comment]
Name "Get and use JWT" activity as a "challenge". Later after merging u can open an issue for ppl to solve that challenge using any backend lang

**Comment by DesignrKnight** [comment]
Done the suggested changes

**Comment by Kevinpaulose05** [comment]
> > Proposed Plan of Change
> > 
> > * [x]   Adding pre-requisite
> > * [x]   Removing long theory section and make it interactive
> > * [x]   A live demo of client-side getting token and using it with API call. Will add at end of the activity, before the quiz
> > * [x]   Additional Info on hash/digital sign
> > * [ ]   Storing JWT and transmitting JWT. Best practices.
> > * [x]   I'll try to make a server-side demo of JWT creation. Will add this also in the demo section with client-side demo
> > 
> > Does this sound good @Ak-Shaw @Kevinpaulose05
> 
> Have done almost all the changes. Have not done JWT storing and transmitting section as that in itself is a vast topic

Do this incomplete part. I guess it includes some real time applications too( like I said google apis or anything u know).

Now to make ur microbyte attractive:
1. Rename ur MB to something that has "JWT" instead of current "jwt". give a proper professional name. This shud be done before us merging.
2. add images for visual appeal. Like some access token, photo of a key, anything funny but related to this( pertaining to "secret", "password", "key", etc.). This u do at ur own pace. Maybe after merging also. Better u do it before.

Don't worry, ur MB is pretty much ready now. Just finish it nice and good

**Comment by DesignrKnight** [comment]
The reason why I did not do Google APIs is that although it uses JWT, it works on OAuth 2.0 that is not just one token but uses
Access Token and Refresh Token along with Client Secret.
If you know any popular API that uses just a basic JWT auth, do tell me

**Comment by Kevinpaulose05** [comment]
One more thing i feel u should add is how jwt used in an application. Like a nodejs snip of requiring the package from npm, like jwt = require("json..."). So that ppl will have an idea that it is something we are doing while developing an  application, a visual ease

**Comment by Kevinpaulose05** [comment]
> The reason why I did not do Google APIs is that although it uses JWT, it works on OAuth 2.0 that is not just one token but uses
> Access Token and Refresh Token along with Client Secret.
> If you know any popular API that uses just a basic JWT auth, do tell me

That is there. Today OAuth is more commonly used actually ;). I said goggle apis from the top of m head.

**Comment by DesignrKnight** [comment]
I did use const jwt = require('jsonwebtoken');  in server side code
Do you want me to show an example of client side app as well?

**Comment by Kevinpaulose05** [comment]
> I did use const jwt = require('jsonwebtoken'); in server side code
> Do you want me to show an example of client side app as well?

Where is it in the readme? If u cud specify, I don;t seem to find it

**Comment by Kevinpaulose05** [comment]
> The reason why I did not do Google APIs is that although it uses JWT, it works on OAuth 2.0 that is not just one token but uses
> Access Token and Refresh Token along with Client Secret.
> If you know any popular API that uses just a basic JWT auth, do tell me

Mention this in ur readme. That jwt is used for small scale basic apps today. OAuth is used in production level apps. But to know that we need to know jwt first. this is a very good point

**Comment by Kevinpaulose05** [comment]
y r u using hyphen in ur name? Remove it, the names good ;)

**Comment by DesignrKnight** [comment]
> > I did use const jwt = require('jsonwebtoken'); in server side code
> > Do you want me to show an example of client side app as well?
> 
> Where is it in the readme? If u cud specify, I don;t seem to find it

Challenge Activities
They both require learners to fetch and use code from `src/js-server`

**Comment by DesignrKnight** [comment]
> y r u using hyphen in ur name? Remove it, the names good ;)


Should I use folder name with spaces instead of the hyphen?

**Comment by Kevinpaulose05** [comment]
> folder name with spaces instead of the hyphen?

Yes

**Comment by Kevinpaulose05** [comment]
> > > I did use const jwt = require('jsonwebtoken'); in server side code
> > > Do you want me to show an example of client side app as well?
> > 
> > 
> > Where is it in the readme? If u cud specify, I don;t seem to find it
> 
> Challenge Activities
> They both require learners to fetch and use code from `src/js-server`

Okay. I entirely missed that part. Now mention basic knowledge of nodejs and npm in prerequisite(for the challenge mostly) and remove CRUD operations' main code from the files. Only then contributors can contribute to ur challenge through the issues u will open

**Comment by DesignrKnight** [comment]
> > > > I did use const jwt = require('jsonwebtoken'); in server side code
> > > > Do you want me to show an example of client side app as well?
> > > 
> > > 
> > > Where is it in the readme? If u cud specify, I don;t seem to find it
> > 
> > 
> > Challenge Activities
> > They both require learners to fetch and use code from `src/js-server`
> 
> Okay. I entirely missed that part. Now mention basic knowledge of nodejs and npm in prerequisite and remove CRUD operations' main code from the files. Only then contributors can contribute to ur challenge through the issues u will open

The code just has basic structure which will work only after changes are done. If I remove more parts of code, then understanding it will become difficult

**Comment by Kevinpaulose05** [comment]
> > > > > I did use const jwt = require('jsonwebtoken'); in server side code
> > > > > Do you want me to show an example of client side app as well?
> > > > 
> > > > 
> > > > Where is it in the readme? If u cud specify, I don;t seem to find it
> > > 
> > > 
> > > Challenge Activities
> > > They both require learners to fetch and use code from `src/js-server`
> > 
> > 
> > Okay. I entirely missed that part. Now mention basic knowledge of nodejs and npm in prerequisite and remove CRUD operations' main code from the files. Only then contributors can contribute to ur challenge through the issues u will open
> 
> The code just has basic structure which will work only after changes are done. If I remove more parts of code, then understanding it will become difficult

True true. I just looked at it in detail now. My bad

**Comment by Kevinpaulose05** [comment]
So now u make all the pending changes and let me know when ready

**Comment by Kevinpaulose05** [comment]
Mention Node.js instead of Nodejs. Give the official site link https://nodejs.org/en/ too

**Comment by DesignrKnight** [comment]
All changes are done. I suppose so

**Comment by Kevinpaulose05** [comment]
If u can reduce the size of the second image it wud be better. Rest all is good

**Comment by DesignrKnight** [comment]
Done

**Review by Kevinpaulose05 — APPROVED** [review]
Apologies for my delay. Awesome work @DesignrKnight . The reason I stretched it was I wanted this topic's MB to be pretty good and u have achieved it, thanks to ur cooperation.

Two things to do now:
1. Open issues for learners to contribute to by following the proper format of issue(given in .github folder of master repo).
2. Showcase ur microbyte on slack microbyte showcase channel using [this](https://github.com/Crio-Bytes/Demo-Repo/blob/master/Maintainer%20Resources/new-microbyte-showcase-template.md)

Happy Contributing!!!

**Comment by Kevinpaulose05** [comment]
@all-contributors kindly add @deepak-prajapatii for:

- [x] code
- [x] doc
- [x] ideas
- [x] content

**Comment by Kevinpaulose05** [comment]
@all-contributors kindly add @DesignrKnight  for:

- [x] code
- [x] doc
- [x] ideas
- [x] content

**Comment by DesignrKnight** [comment]
Thank you so much for all the feedback and fast review

**Pull request merged**
